### ISSUE TITLE
Cannot pull from mutlilangual site

### ISSUE THREAD

**Issue by akemrir** [locomotivecms/wagon#336]
```# Error description:                                                   
undefined method `attributes' for nil:NilClass                         

# Backtrace:                                                           
        /home/akemrir/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/locomotivecms_wagon-2.1.1/lib/locomotive/wagon/commands/pull_sub_commands/pull_content_entries_command.rb:85:in `value_of'                                                                                             
        /home/akemrir/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/locomotivecms_wagon-2.1.1/lib/locomotive/wagon/commands/pull_sub_commands/pull_content_entries_command.rb:37:in `block (3 levels) in yaml_attributes'                                                                  
        /home/akemrir/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/locomotivecms_wagon-2.1.1/lib/locomotive/wagon/commands/pull_sub_commands/pull_content_entries_command.rb:37:in `each'
....
```

This situation happens on wagon v2.1.1 and cms 3.0.1
And with wagon 2.2.0 and cms 3.2

local site.yml have locales set to [de, en], same as target sync site.
Changing order doesn't help.

I was debugging it, and searching for object in database on which error occured.
It happend for name _slug
But in production database this entry have all of this keys. Some with empty strings but all translated attributes have both keys.
I have tried to edit this entry and save again, but with same result.

Is there a fix for it?

**Comment by did** [comment]
hi @akemrir, it looks new to me. Send me a private message on Gitter: https://gitter.im/locomotivecms/wagon?utm_source=share-link&utm_medium=link&utm_campaign=share-link. My nickname is did.

**Comment by akemrir** [comment]
hi, I wrote to you on Gitter
you was on vacation?

**Comment by clecoufl** [comment]
Hi @did ,

I have encountered the same issue with a multilingual site.
A little bit of history : 
1. Website was created in fr (first default language)
2. Some content_types were added (with some instances)
3. Then we added a second language (en)
4. Added some more content_types and instances
5. Switch default language to en
6. Continued building the website with a lot of new data in the content_types

Engine version : 3.2.1

At a certain point wagon pull started failing with the above error.
We pulled the latest github version of wagon and started investigating.

We tracked down the issue to lib/locomotive/wagon/commands/pull_sub_commands/pull_content_entries_command.rb
in the method : def fetch_content_entries(content_type, &block)
Original code :
```
  def fetch_content_entries(content_type, &block)
      page = 1
      while page do
        entries, _next_page = {}, nil

        locales.each do |locale|
          next if locale != default_locale && content_type.localized_names.empty?

          (_entries = api_client.content_entries(content_type).all(nil, { page: page }, locale)).each do |entry|
            (entries[entry._id] ||= {})[locale] = entry
          end

          _next_page = _entries._next_page if _next_page.nil?
        end

        yield(entries.values)

        page = _next_page
      end
    end
```

With debug outputs, we discovered that some entries were not available in both website locales (some only in fr, others only in en).
In my understanding of the original code, entries are retrieved by batches of 10 elements in the default language, then for each active language the same page of entries is fetched and the entries in each language are consolidated into the entries loop variable (by entry id).
ID1 => {'en' => {something}, 'fr' => {something}  } etc..

However we noticed that the consolidated entries variable was missing languages for some given IDs : 
ID_XX => {'en' => {something}, 'fr' => {something}  } 
....
ID_XY => {'fr' => {something}  } 
ID_ZZ => {'en' => {something} } 

We believe that the order of the elements received by the method : 
api_client.content_entries(content_type).all 
is not exactly the same for each language : 
ex : EN pull IDs (1, 2, 3 ,4 ,5 ,6 ,7 ,8, 9, 10), FR pull IDs (1, 2, 3, 4, 5, 6, 8, 9, 10, 11).

As a temporary fix we changed the code to build the entries map on all pages before calling the yield block: 
```
    def fetch_content_entries(content_type, &block)
      page = 1
      entries = {}
      while page do
        _next_page = nil

        locales.each do |locale|
          next if locale != default_locale && content_type.localized_names.empty?

          (_entries = api_client.content_entries(content_type).all(nil, { page: page }, locale)).each do |entry|
            (entries[entry._id] ||= {})[locale] = entry
          end

              _next_page = _entries._next_page if _next_page.nil?
        end

        page = _next_page
      end

      yield(entries.values)

    end
```

Which fixed the issue (however it has some other undesired side effects when pulling a lot of content, wagon will now eventually - not always - trigger a Faraday timeout error during the pull).

A real fix would imply adapting the ordering of the elements in the query used by : 
api_client.content_entries(content_type).all(nil, { page: page }, locale))
to have a deterministic order of elements by page and locale.

Hope this helps ;)
